firebase-app-distribution/src/test/java/com/google/firebase/appdistribution/UpdateApkClientTest.java

```diff
@@ -0,0 +1,169 @@
+
+  private UpdateApkClient updateApkClient;
+  private TestActivity activity;
+  private ShadowActivity shadowActivity;
+  private CountDownLatch latch;
+  @Mock private File mockFile;
+
+  static class TestActivity extends Activity {}
+
+  @Before
+  public void setup() {
+
+    MockitoAnnotations.initMocks(this);
+
+    FirebaseApp.clearInstancesForTest();
+
+    FirebaseApp firebaseApp =
+        FirebaseApp.initializeApp(
+            ApplicationProvider.getApplicationContext(),
+            new FirebaseOptions.Builder()
+                .setApplicationId(TEST_APP_ID_1)
+                .setProjectId(TEST_PROJECT_ID)
+                .setApiKey(TEST_API_KEY)
+                .build());
+
+    activity = Robolectric.buildActivity(TestActivity.class).create().get();
+    shadowActivity = shadowOf(activity);
+
+    this.latch = new CountDownLatch(1);
+    when(mockFile.getPath()).thenReturn(TEST_URL);
+    when(mockFile.length()).thenReturn(TEST_FILE_LENGTH);
+
+    this.updateApkClient = Mockito.spy(new UpdateApkClient(firebaseApp));
+  }
+
+  @Test
+  public void updateApk_whenDownloadFails_setsNetworkError() throws Exception {
+    UpdateTaskImpl updateTask = new UpdateTaskImpl();
+    updateApkClient.updateApk(updateTask, TEST_URL, activity);
+    latch.await(1000, TimeUnit.MILLISECONDS);
+    assertFalse(updateTask.isSuccessful());
+    assertTrue(updateTask.getException() instanceof FirebaseAppDistributionException);
+    FirebaseAppDistributionException e =
+        (FirebaseAppDistributionException) updateTask.getException();
+    assertEquals(Constants.ErrorMessages.NETWORK_ERROR, e.getMessage());
+    assertEquals(FirebaseAppDistributionException.Status.DOWNLOAD_FAILURE, e.getErrorCode());
+  }
+
+  @Test
+  public void updateApk_whenInstallSuccessful_setsResult() throws Exception {
+    List<UpdateProgress> progressEvents = new ArrayList<>();
+    UpdateTaskImpl updateTask = new UpdateTaskImpl();
+    doReturn(Tasks.forResult(mockFile)).when(updateApkClient).downloadApk(TEST_URL, updateTask);
+    updateTask.addOnProgressListener(progressEvents::add);
+
+    updateApkClient.updateApk(updateTask, TEST_URL, activity);
+    latch.await(1000, TimeUnit.MILLISECONDS);
+    updateApkClient.setInstallationResult(RESULT_OK);
+    assertEquals(1, progressEvents.size());
+    assertEquals(
+        UpdateProgress.builder()
+            .setApkBytesDownloaded(TEST_FILE_LENGTH)
+            .setApkFileTotalBytes(TEST_FILE_LENGTH)
+            .setUpdateStatus(UpdateStatus.DOWNLOADED)
+            .build(),
+        progressEvents.get(0));
+    assertTrue(updateTask.isSuccessful());
+  }
+
+  @Test
+  public void updateApk_whenInstallCancelled_setsError() throws Exception {
+    UpdateTaskImpl updateTask = new UpdateTaskImpl();
+    doReturn(Tasks.forResult(mockFile)).when(updateApkClient).downloadApk(TEST_URL, updateTask);
+
+    updateApkClient.updateApk(updateTask, TEST_URL, activity);
+    latch.await(1000, TimeUnit.MILLISECONDS);
```

Comment:
Do we call latch.countDown() anywhere? If not, could we try Thread.sleep() for simplicity?

Reply:
Yep Thread.sleep() works and I don't countdown, going to change it.